src/modules/formBuilder/components/FormBuilderPage.tsx

```diff
@@ -1,86 +1,116 @@
 import Loading from '@/components/elements/Loading';
+import NotFound from '@/components/pages/NotFound';
 import useSafeParams from '@/hooks/useSafeParams';
-import FormBuilderHeader from '@/modules/formBuilder/components/FormBuilderHeader';
-import FormBuilderPalette from '@/modules/formBuilder/components/FormBuilderPalette';
-import FormItemEditor from '@/modules/formBuilder/components/FormItemEditor';
-import FormTree from '@/modules/formBuilder/components/formTree/FormTree';
+import {
+  emptyErrorState,
+  ErrorState,
+  partitionValidations,
+} from '@/modules/errors/util';
+import FormBuilder from '@/modules/formBuilder/components/FormBuilder';
 import { formatDateForDisplay } from '@/modules/hmis/hmisUtil';
 import {
+  FormDefinitionJson,
   FormItem,
   useGetFormDefinitionFieldsForEditorQuery,
+  useUpdateFormDefinitionMutation,
 } from '@/types/gqlTypes';
 
 const FormBuilderPage = () => {
   const { formId } = useSafeParams() as { formId: string };
 
-  const { data: { formDefinition } = {}, error } =
-    useGetFormDefinitionFieldsForEditorQuery({
-      variables: { id: formId },
-    });
+  const [workingDefinition, setWorkingDefinition] = useState<
+    FormDefinitionJson | undefined
+  >();
+
+  const {
+    data: { formDefinition: initialFormDefinition } = {},
+    loading: fetchLoading,
+    error: fetchError,
+  } = useGetFormDefinitionFieldsForEditorQuery({
+    variables: { id: formId },
+    onCompleted: (data) => {
+      setWorkingDefinition(data.formDefinition?.definition);
+    },
+  });
 
-  // TODO - update the API to return correct values
+  // TODO(#6090) - update the API to return correct values
   const lastUpdatedDate = formatDateForDisplay(new Date());
   const lastUpdatedBy = 'User Name';
 
   const [selectedItem, setSelectedItem] = useState<FormItem | undefined>(
     undefined
   );
+  const [originalLinkId, setOriginalLinkId] = useState<string | undefined>(
+    undefined
+  );
+
+  const closeItemEditor = () => {
+    setSelectedItem(undefined);
+    setOriginalLinkId(undefined);
+  };
+
+  const [errorState, setErrorState] = useState<ErrorState>(emptyErrorState);
+
+  const [
+    updateFormDefinition,
+    { loading: saveLoading, error: saveError, data },
+  ] = useUpdateFormDefinitionMutation({
+    onCompleted: (data) => {
+      if (
+        data.updateFormDefinition?.errors &&
+        data.updateFormDefinition.errors.length > 0
+      ) {
+        setErrorState(partitionValidations(data.updateFormDefinition.errors));
+      } else {
+        setWorkingDefinition(
+          data.updateFormDefinition?.formDefinition?.definition
+        );
+        setErrorState(emptyErrorState);
+        closeItemEditor();
+      }
+    },
+  });
+
+  const formDefinition = useMemo(() => {
+    if (data && data.updateFormDefinition?.formDefinition)
+      return data.updateFormDefinition.formDefinition;
+    return initialFormDefinition;
+  }, [data, initialFormDefinition]);
```

Comment:
so I could see this causing bugs in the event that the `updateFormDefinition` mutation _failed_, then `data` would be present but `data.updateFormDefinition?.formDefinition` would be nil. I don't think it is correct to use `initialFormDefinition` in that scenario. But I'm not sure what it should be without tinkering with error scenarios.

Reply:
That's interesting. I think in that scenario it should be the most recent successfully saved version. Which is `initialFormDefinition` if the user has not already made and saved changes on this page. But if they have, then it's the last result of the update mutation, which gets overridden when the failed mutation returns.

Reply:
ah right, good point. that seems fine to me.